### PR TITLE
[Serializer] Improve performance by exposing supports-never/-always 

### PR DESCRIPTION
**Pull request by Jeroeny**
| Q             | A
| ------------- | ---
| Branch?       | 6.3
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | Likely yes
| License       | MIT
| Doc PR        | Not yet initiated

**Description**

In the Serializer component, there is a `CacheableSupportsMethodInterface`. It can indicate the fact the normalizer's `supports*` method result can be cached.
Unfortunately, every normalizer that determines this result using `$context` cannot have its result cached (because it's based on `format`+`type`).

However there is some information with that supports call that often can be cached. Which is when a normalizer _never_ supports a `format`+`type`, regardless of context. 

For example:

```php
public function supportsNormalization(mixed $data, string $format = null /*, array $context = [] */)
{
    return $data instanceof DateTime && ($context['DB_TYPES') ?? false);
}
```

This normalizer would not be able to have its result cached. Even though we know that if the `$data` is not an instance of `DateTime`, it will _never_  be supported, regardless of `$context`.

**Solution(s)**

This pull request proposes to introduce the ability to return `always/never supports $format+$type`, to increase performance.  
By introducing or altering a method to return one of 4 values: 

- support never
- support not
- support
- support always

There are multiple ways to achieve this. 

1. The one that is done in this PR for now, is to use the existing `supportsNormalization` method.
It would be changed to allow it to return one of the enum values (e.g. using integers -1 to 2, or bitwise), while still allowing `bool` return for backwards compatibilty. 
This has the advantage of only needing one method for the caching and supports information. And no extra interfaces, so `CacheableSupportsMethodInterface` can be deprecated.

With this, the example from earlier would become:

```php
public function supportsNormalization(mixed $data, string $format = null /*, array $context = [] */)
{
    if (! $data instanceof DateTime) {
        return SUPPORT_NEVER; // a const, or could use an enum/integer directly
    }

    return ($context['DB_TYPES') ?? false) ? SUPPORT : SUPPORT_NOT;
}
```

2. A new method `CacheableSupportsInterface::supportsNormalizationCached($data, $format, $context): int`. Possibly with a trait that implements the `supportsNormalization` to prevent duplicate logic (by doing `return $this->supportsNormalizationCached(..) >= SUPPORT`).

The solutions apply to both normalization and denormalization.

**Performance impact example**

- If there are 10 (de)normalizers, that cannot use `CacheableSupportsMethodInterface` because of `$context`, but can return `SUPPORT_NEVER` like the example above. 
- The last of the normalizers supports the `DateTime` type. 

Now you normalize a `$dateTime` 1000 times:

- Before: 10.000 calls to `supportsNormalization()` and iterated over 10.010 `$cached` normalizers .
- After: 1.009 calls to `supportsNormalization()` and iterated over 1.010 `$cached` normalizers (`$this->normalizerCache[$format][DateTime::class]` now only contains the one relevant normalizer).

A decrease in method calls of 89.91%.

### CODE REVIEW

File: src/Symfony/Component/Serializer/Normalizer/CacheableSupportsInterface.php [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    // -1 : no, never supports the $format+$type, cache it
+    // 0  : no, no cache
+    // 1  : yes, no cache
+    // 2  : yes, always supports the $format+$type, cache it
+
+    public const SUPPORT_NEVER = -1;
+    public const SUPPORT_NOT = 0;
+    public const SUPPORT = 1;
+    public const SUPPORT_ALWAYS = 2;
+}
```
Comment (Jeroeny) [inline src/Symfony/Component/Serializer/Normalizer/CacheableSupportsInterface.php]: Alternatively, make the interface extend `interface CacheableSupportsInterface extends NormalizerInterface` to narrow the return type.

```suggestion
/** @return int **/
public function supportsNormalization(mixed $data, string $format = null /*, array $context = [] */);

}
```

File: src/Symfony/Component/Serializer/Serializer.php [hunk 2]
```diff
@@ -257,34 +258,38 @@ public function supportsDenormalization(mixed $data, string $type, string $forma
...
-                if (!$normalizer instanceof CacheableSupportsMethodInterface || !$normalizer->hasCacheableSupportsMethod()) {
-                    $this->normalizerCache[$format][$type][$k] = false;
-                } elseif ($normalizer->supportsNormalization($data, $format, $context)) {
-                    $this->normalizerCache[$format][$type][$k] = true;
-                    break;
+                $cacheable = $normalizer instanceof CacheableSupportsInterface;
+                if (!$cacheable && (!$normalizer instanceof CacheableSupportsMethodInterface || !$normalizer->hasCacheableSupportsMethod())) {
+                    $this->normalizerCache[$format][$type][$k] = CacheableSupportsInterface::SUPPORT_NOT;
+                } elseif (($support = $normalizer->supportsNormalization($data, $format, $context)) || ($cacheable && $support > CacheableSupportsInterface::SUPPORT_NEVER)) {
+                    $support = $cacheable ? $support : ($support ? CacheableSupportsInterface::SUPPORT_ALWAYS : CacheableSupportsInterface::SUPPORT_NEVER);
```
Comment (Jeroeny) [inline src/Symfony/Component/Serializer/Serializer.php]: Alternatively (this approach would remove the need for the `CacheableSupportsInterface`):

```suggestion
                $support = $normalizer->supportsNormalization($data, $format, $context);
                if (! is_int($support)) {
                    if ($normalizer instanceof CacheableSupportsMethodInterface && $normalizer->hasCacheableSupportsMethod()) {
                        $support = $support ? CacheableSupportsInterface::SUPPORT_ALWAYS : CacheableSupportsInterface::SUPPORT_NEVER;
                    } else {
                        $support = $support ? CacheableSupportsInterface::SUPPORT : CacheableSupportsInterface::SUPPORT_NOT;
                    }
                }

                if ($support === CacheableSupportsInterface::SUPPORT_NEVER) {
                    continue;
                }
```

**Comment by Jeroeny** [comment]
Is release 6.1 feature-locked yet, or could this still be considered for it ?

**Comment by Jeroeny** [comment]
@dunglas Github has autosuggested you as reviewer. Do you think that's correct or would there be a better fit that I could contact for a review? Thanks

File: src/Symfony/Component/Serializer/Normalizer/CacheableSupportsInterface.php [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+ * @author Jeroen Spee <https://github.com/Jeroeny>
+ */
+interface CacheableSupportsInterface
+{
+    // -1 : no, never supports the $format+$type, cache it
+    // 0  : no, no cache
+    // 1  : yes, no cache
+    // 2  : yes, always supports the $format+$type, cache it
+
+    public const SUPPORT_NEVER = -1;
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/CacheableSupportsInterface.php]: This could maybe be an enum?

File: src/Symfony/Component/Serializer/Serializer.php [hunk 4]
```diff
@@ -257,34 +258,38 @@ public function supportsDenormalization(mixed $data, string $type, string $forma
     /**
      * Returns a matching normalizer.
      *
-     * @param mixed  $data    Data to get the serializer for
-     * @param string $format  Format name, present to give the option to normalizers to act differently based on formats
-     * @param array  $context Options available to the normalizer
+     * @param mixed   $data    Data to get the serializer for
+     * @param ?string $format  Format name, present to give the option to normalizers to act differently based on formats
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Serializer.php]: This should probably be done in another PR, but good catch!

File: src/Symfony/Component/Serializer/Normalizer/CacheableSupportsInterface.php [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+/**
+ * Marker interface for normalizers and denormalizers that return
+ * an integer in their supports*() methods.
+ *
+ * By implementing this interface, the return value of the
+ * supports*() methods will be cached by type and format, based on the returned integer.
+ *
+ * @author Jeroen Spee <https://github.com/Jeroeny>
+ */
+interface CacheableSupportsInterface
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/CacheableSupportsInterface.php]: Couldn't we use the existing interface instead and deprecate returning a bool?

**Review by dunglas — CHANGES_REQUESTED** [review]
Some nitpicking, otherwise LGTM.

File: src/Symfony/Component/Serializer/Serializer.php (hunk 4 above)
Comment (Jeroeny) [reply src/Symfony/Component/Serializer/Serializer.php]: True, should I extract it to another PR?

File: src/Symfony/Component/Serializer/Normalizer/CacheableSupportsInterface.php (hunk 3 above)
Comment (Jeroeny) [reply src/Symfony/Component/Serializer/Normalizer/CacheableSupportsInterface.php]: That'd be nice if possible. I've made it an enum. Is the casing `SupportNever` correct for Enums in Symfony ?

File: src/Symfony/Component/Serializer/Normalizer/CacheableSupportsInterface.php (hunk 5 above)
Comment (Jeroeny) [reply src/Symfony/Component/Serializer/Normalizer/CacheableSupportsInterface.php]: Yes, that's also a possibility and makes adopting the new return type possibly easier. I've removed this interface.

File: src/Symfony/Component/Serializer/Normalizer/CacheableSupport.php [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Serializer\Normalizer;
+
+/**
```
Comment (Jeroeny) [inline src/Symfony/Component/Serializer/Normalizer/CacheableSupport.php:14]: The codestyle claims this should be a single-asterisk docblock (`/*`), even though classes and interfaces do have `/**`, is the check perhaps not yet updated to support enums?

```suggestion
/*
```

Comment (dunglas) [reply src/Symfony/Component/Serializer/Normalizer/CacheableSupport.php:14]: imho it's a bug in fabbot's config

**Comment by chalasr** [comment]
Can you rebase and add a changelog entry?

**Comment by chalasr** [comment]
Actually I'd prefer to let us more time to review this change as I'm not fond of the proposed API. 
Postponing to 6.2 if you don't mind. Thanks

**Comment by Jeroeny** [comment]
@chalasr Alright, do you have any suggestion for the implementation ?

**Comment by chalasr** [comment]
@Jeroeny not yet, we're focus on stabilizing 6.1 right now. But I'm confident giving us some more time will allow us to explore other possibilities, even if we stick to this one in the end.

**Comment by Jeroeny** [comment]
Rebased and targeting `6.2`. Added changelog entry (even though it may still change).

File: src/Symfony/Component/Serializer/Normalizer/CacheableSupport.php [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+ * 2  : yes, always supports the $format+$type, cache it
+ *
+ * @author Jeroen Spee <https://github.com/Jeroeny>
+ */
+enum CacheableSupport: int
+{
+    case SupportNever = -1;
+    case SupportNot = 0;
+    case Support = 1;
+    case SupportAlways = 2;
```
Comment (mtarld) [inline src/Symfony/Component/Serializer/Normalizer/CacheableSupport.php:30]: What about adding a `support` method here?
```php
public function support(): bool
{
    return match($this) {
        CacheableSupport::SupportNever, CacheableSupport::SupportNot => false,
        CacheableSupport::Support, CacheableSupport::SupportAlways => true,
    };
}
```

File: src/Symfony/Component/Serializer/Serializer.php [hunk 8]
```diff
@@ -264,34 +265,48 @@ public function supportsDenormalization(mixed $data, string $type, string $forma
...
+                if (CacheableSupport::SupportAlways === $support) {
                     break;
                 }
             }
         }
 
         foreach ($this->normalizerCache[$format][$type] as $k => $cached) {
             $normalizer = $this->normalizers[$k];
-            if ($cached || $normalizer->supportsNormalization($data, $format, $context)) {
+            if ($cached->value >= $minCached->value || ($cached->value > $minUncached->value && $this->supportsNormalizationWrapper($normalizer, $data, $format, $context)->value > CacheableSupport::SupportNot->value)) {
```
Comment (mtarld) [inline src/Symfony/Component/Serializer/Serializer.php:276]: Just wondering, why is it different from the `getDenormalizer`? (I'm talking about the `minUncached` for example)
```
            if ($cached->value >= $minCached->value || ($cached->value > CacheableSupport::SupportNever->value && $this->supportsDenormalizationWrapper($normalizer, $data, $class, $format, $context)->value > CacheableSupport::SupportNot->value)) {
```

File: src/Symfony/Component/Serializer/Normalizer/CacheableSupport.php (hunk 7 above)
Comment (Jeroeny) [reply src/Symfony/Component/Serializer/Normalizer/CacheableSupport.php:30]: Makes sense, added

File: src/Symfony/Component/Serializer/Serializer.php (hunk 8 above)
Comment (Jeroeny) [reply src/Symfony/Component/Serializer/Serializer.php:276]: I think they can be the same indeed

File: src/Symfony/Component/Serializer/Normalizer/CacheableSupportsMethodInterface.php [hunk 9]
```diff
@@ -19,6 +19,8 @@
  * supports*() methods will be cached by type and format.
  *
  * @author Kévin Dunglas <dunglas@gmail.com>
+ *
+ * @deprecated since symfony/serializer 6.1, return CacheableSupport from the supports*() method instead
```
Comment (mtarld) [inline src/Symfony/Component/Serializer/Normalizer/CacheableSupportsMethodInterface.php:23]: As a new deprecation is added, you must add an entry in the `UPGRADE-6.2.md` file and add another entry mentioning that deprecation in the `CHANGELOG.md` file

File: src/Symfony/Component/Serializer/Normalizer/CacheableSupport.php [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Serializer\Normalizer;
+
+/**
+ * The return value of supports*() methods  in {@see NormalizerInterface} will be cached by type and format, based on the returned SUPPORT_* value.
```
Comment (mtarld) [inline src/Symfony/Component/Serializer/Normalizer/CacheableSupport.php]:
```suggestion
 * The return value of supports*() methods in {@see NormalizerInterface} and {@see DenormalizerInterface}. 
 * Tells if the result should be cached based on type and format.
```
WDYT?

File: src/Symfony/Component/Serializer/Serializer.php [hunk 11]
```diff
@@ -247,34 +248,48 @@ public function supportsDenormalization(mixed $data, string $type, string $forma
...
         }
 
         return null;
     }
 
+    private function supportsNormalizationWrapper(NormalizerInterface $normalizer, mixed $data, ?string $format, array $context): CacheableSupport
+    {
+        return match ($value = $normalizer->supportsNormalization($data, $format, $context)) {
+            true => $normalizer instanceof CacheableSupportsMethodInterface && $normalizer->hasCacheableSupportsMethod() ? CacheableSupport::SupportAlways : CacheableSupport::Support,
+            false => $normalizer instanceof CacheableSupportsMethodInterface && $normalizer->hasCacheableSupportsMethod() ? CacheableSupport::SupportNever : CacheableSupport::SupportNot,
```
Comment (mtarld) [inline src/Symfony/Component/Serializer/Serializer.php:296]: I think that we might trigger a deprecation there if `$data` is a boolean, saying that normalizers must return a `CacheableSupport` instead.

File: src/Symfony/Component/Serializer/CHANGELOG.md [hunk 12]
```diff
@@ -11,6 +11,7 @@ CHANGELOG
  * Change the signature of `AttributeMetadataInterface::setSerializedName()` to `setSerializedName(?string)`
  * Change the signature of `ClassMetadataInterface::setClassDiscriminatorMapping()` to `setClassDiscriminatorMapping(?ClassDiscriminatorMapping)`
  * Add option YamlEncoder::YAML_INDENTATION to YamlEncoder constructor options to configure additional indentation for each level of nesting. This allows configuring indentation in the service configuration.
+ * Improve serializer performance by exposing supports-never/-always in normalizers
```
Comment (mtarld) [inline src/Symfony/Component/Serializer/CHANGELOG.md]: IMHO, this might be more "developer-oriented", what do you think about something like:
```md
* Cache normalizer selection based on format and type
```
?

File: src/Symfony/Component/Serializer/Serializer.php [hunk 13]
```diff
@@ -247,34 +248,48 @@ public function supportsDenormalization(mixed $data, string $type, string $forma
...
+                if (CacheableSupport::SupportAlways === $support) {
                     break;
                 }
             }
         }
 
         foreach ($this->normalizerCache[$format][$type] as $k => $cached) {
             $normalizer = $this->normalizers[$k];
-            if ($cached || $normalizer->supportsNormalization($data, $format, $context)) {
+            if ($cached->value >= $minCached->value || ($cached->value > $minUncached->value && $this->supportsNormalizationWrapper($normalizer, $data, $format, $context)->value > CacheableSupport::SupportNot->value)) {
```
Comment (mtarld) [inline src/Symfony/Component/Serializer/Serializer.php:276]: Not sure if it work like that, but WDYT about this approach instead?
```php
        $type = \is_object($data) ? $data::class : 'native-'.\gettype($data);

        if (!isset($this->normalizerCache[$format][$type])) {
            $this->normalizerCache[$format][$type] = [];

            foreach ($this->normalizers as $k => $normalizer) {
                if (!$normalizer instanceof NormalizerInterface) {
                    continue;
                }

                $support = $this->supportsNormalizationWrapper($normalizer, $data, $format, $context);
                if (CacheableSupport::SupportNever === $support) {
                    continue;
                }

                $this->normalizerCache[$format][$type][$k] = $support;
                if (CacheableSupport::SupportAlways === $support) {
                    break;
                }
            }
        }

        foreach ($this->normalizerCache[$format][$type] as $k => $cached) {
            $normalizer = $this->normalizers[$k];
            if ($cached->supports() || $this->supportsNormalizationWrapper($normalizer, $data, $format, $context)->supports()) {
                return $normalizer;
            }
        }

        return null;
```
This seems more straightforward to me

File: src/Symfony/Component/Serializer/Serializer.php [hunk 14]
```diff
@@ -247,34 +248,48 @@ public function supportsDenormalization(mixed $data, string $type, string $forma
...
                 return $normalizer;
             }
         }
 
         return null;
     }
 
+    private function supportsNormalizationWrapper(NormalizerInterface $normalizer, mixed $data, ?string $format, array $context): CacheableSupport
```
Comment (mtarld) [inline src/Symfony/Component/Serializer/Serializer.php:287]: Maybe we can find a more meaningful name for that private method?
Or at least add a comment telling that is a BC layer?

File: src/Symfony/Component/Serializer/Tests/SerializerTest.php [hunk 15]
```diff
@@ -1247,6 +1248,29 @@ public function provideCollectDenormalizationErrors()
...
+
+                return ($context['TEST_CONTEXT'] ?? false) ? CacheableSupport::Support : CacheableSupport::SupportNot;
+            });
+
+        $this->assertTrue($serializer->supportsNormalization(new Bar(''), 'json', ['TEST_CONTEXT' => true]));
+        $this->assertFalse($serializer->supportsNormalization(new Bar(''), 'json'));
+        $this->assertFalse($serializer->supportsNormalization(new \StdClass(), 'json', ['TEST_CONTEXT' => true]));
+        $this->assertFalse($serializer->supportsNormalization(new \StdClass(), 'json', ['TEST_CONTEXT' => true]));
+        $this->assertFalse($serializer->supportsNormalization(new \StdClass(), 'json', ['TEST_CONTEXT' => true]));
+    }
```
Comment (mtarld) [inline src/Symfony/Component/Serializer/Tests/SerializerTest.php:1273]: Maybe we can test the `SupportNever` as well?

**Review by mtarld — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Serializer/Normalizer/CacheableSupportsMethodInterface.php (hunk 9 above)
Comment (Jeroeny) [reply src/Symfony/Component/Serializer/Normalizer/CacheableSupportsMethodInterface.php:23]: Thanks, added

File: src/Symfony/Component/Serializer/Normalizer/CacheableSupport.php (hunk 10 above)
Comment (Jeroeny) [reply src/Symfony/Component/Serializer/Normalizer/CacheableSupport.php]: Looks good to me (I now see the 'commit suggestion' button, should've used that..)

File: src/Symfony/Component/Serializer/Serializer.php (hunk 14 above)
Comment (Jeroeny) [reply src/Symfony/Component/Serializer/Serializer.php:287]: Added a comment. The name could be better indeed, but haven't thought of anything better yet, am open to suggestions.

File: src/Symfony/Component/Serializer/Tests/SerializerTest.php (hunk 15 above)
Comment (Jeroeny) [reply src/Symfony/Component/Serializer/Tests/SerializerTest.php:1273]: This is testing that, by calling `supports..`  multiple times and ensuring `supportsNormalization` is only called 3 times instead of 5. 

```
$normalizer
            ->expects($this->exactly(3))
            ->method('supportsNormalization')
```

Unless you'd say something more explicit could be added?

File: src/Symfony/Component/Serializer/Serializer.php (hunk 13 above)
Comment (Jeroeny) [reply src/Symfony/Component/Serializer/Serializer.php:276]: That is actually not logically the same. The `minUncached` is a variable, because if the cache is fresh, the comparison can be made using the `SupportNot` and `Support` values. Because the evaluation was already done with the `$context` from that call. 

If the cache was not fresh, it must have `SupportAlways`, otherwise evaluate the part after the `||` (which first checks if it can support at all by checking if it doesn't have `SupportNever`).

File: src/Symfony/Component/Serializer/Serializer.php (hunk 11 above)
Comment (Jeroeny) [reply src/Symfony/Component/Serializer/Serializer.php:296]: Indeed, added

**Comment by nicolas-grekas** [comment]
Thanks for giving this a try. I think the approach in #49291 might provide better results.

**Comment by nicolas-grekas** [comment]
Closing in favor #49291, thanks again.
